mabwiser/linear.py

```diff
@@ -162,10 +193,14 @@ def _predict_contexts(self, contexts: np.ndarray, is_predict: bool,
         # Create an empty list of predictions
         predictions = [None] * len(contexts)
         for index, row in enumerate(contexts):
+            rng = np.random.RandomState(seed=seeds[index])
```

Comment:
This line needs to be commented well. 

Spelling out the control flow for learning purposes.. So this class itself has a "self.rng"... but we don't use that here.. and instead, look at the "seeds" which is given as input.

Then, we create a *new* rng based on the particular seed... we then copy that to all arms.. 

Previous _Linears, linUCB and Ridge did not need that BUT LinTS needs it?

Should these only happen for LinTS? 

CC @bkleyn

Reply:
Ridge and LinUCB don't have any random behavior

Reply:
Previously _RidgeRegression and _LinUCB had their own rng objects, but it wasn't used for anything. (I'm not exactly sure why this was the case given that these classes do not inherit from BaseMAB?) But yes, now we create a new rng for each context using seeds that were generated from the "master" rng held by BaseMAB. 

I don't think the fact that the _RidgeRegression and _LinUCB classes do not use the rng object should be an issue. The same is true for the _UCB1 class that has a rng object that's not used in the learning policy.

Reply:
_RidgeRegression and _LinUCB have rng objects because _LinTS was originally intended to be part of our initial release. When we removed it we removed only that class and left _RidgeRegression unchanged. Since neither _RidgeRegression nor _LinUCB uses the rng object, it was left out of the changes made to add parallelization. However _LinTS does use the rng, so we now need to add this change which mirrors how we handle the parallel rngs in the _NearestNeighbors classes.

Reply:
Thanks for the clarification!